pkg/spdk/engine.go

```diff
@@ -1229,6 +1333,399 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+	)
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	defer func() {
+		if retErr != nil {
+			e.log.WithError(engineErr).Errorf("Engine %s failed to expand", e.Name)
+			e.lastExpansionError = errors.Wrap(retErr, "engine failed to expand expansion").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+			}
+			e.ErrorMsg = engineErr.Error()
+
+			if errUpdateLogger := e.log.UpdateLogger(logrus.Fields{
+				"replicaStatusMap": e.ReplicaStatusMap,
+			}); errUpdateLogger != nil {
+				e.log.WithError(errUpdateLogger).Warn("Failed to update logger with replica status map during engine creation")
+			}
+
+			e.lastExpansionError = errors.Wrap(retErr, "engine under non-recoverable operation").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		e.finishExpansion(expanded, size)
+	}()
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		retErr = err
+		return retErr
+	}
+	defer e.closeRplicaClients(replicaClients)
+	if e.isAnyReplicaUnderRebuilding(replicaClients) {
+		retErr = errors.New("expansion failed: at least one replica is under rebuilding")
+		return retErr
+	}
+
+	// prepareRaidForExpansion checks if RAID exists, suspends frontend if needed, and deletes the RAID bdev.
+	isSuspended, bdevRaidUUID, err := e.prepareRaidForExpansion(spdkClient)
+	if isSuspended {
+		defer func() {
+			if frontendErr := e.initiator.Resume(); frontendErr != nil {
+				frontendErr = errors.Wrapf(frontendErr, "failed to resume NVMe initiator during engine %s", e.Name)
+				if engineErr != nil {
+					engineErr = errors.Wrapf(engineErr, "original error; resume failed: %v", frontendErr)
+				} else {
+					engineErr = frontendErr
+				}
+			}
+		}()
+	}
+	if err != nil {
+		engineErr = errors.Wrap(err, "prepare raid for expansion failed")
+		retErr = engineErr
+		return retErr
+	}
+
+	// expand replicas
+	if err := e.expandReplicas(replicaClients, spdkClient, size); err != nil {
+		engineErr = err
+		retErr = err
+		return retErr
+	}
+
+	// recreate RAID and reconnect frontend
+	if err := e.reconnectFrontend(spdkClient, bdevRaidUUID); err != nil {
+		engineErr = errors.Wrap(err, "reconnectFrontend failed")
+		retErr = engineErr
+		return retErr
+	}
+	e.log.Info("Expanding engine complete")
+
+	expanded = true // which could be true even in partial success
+	return nil
+}
+
+func (e *Engine) startExpansion(size uint64) (err error) {
+	if e.isExpanding {
+		return fmt.Errorf("expansion is in progress")
+	}
+
+	if e.IsRestoring {
+		return fmt.Errorf("restoring is in progress")
+	}
+
+	if e.SpecSize > size {
+		return fmt.Errorf("cannot expand engine to a smaller size %v, current size %v", size, e.SpecSize)
+	} else if e.SpecSize == size {
+		e.log.Infof("Replica had been expanded to size %v", size)
```

Comment:
Should we make sure that the caller Expand() also return immediately instead of continue to perform the expansion?